linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -673,19 +676,53 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
   private def transformClosureCommon(arrow: Boolean,
       captureParams: List[ParamDef], params: List[ParamDef],
       restParam: Option[ParamDef], body: Tree,
-      newCaptureValues: List[Tree])(
-      implicit scope: Scope, pos: Position): Closure = {
+      tcaptureValues: List[PreTransform])(cont: PreTransCont)(
+      implicit scope: Scope, pos: Position): TailRec[Tree] = {
 
     val thisType = if (arrow) NoType else AnyType
-    val (allNewParams, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
-    val (newCaptureParams, newParams, newRestParam) = {
-      val (c, t) = allNewParams.splitAt(captureParams.size)
-      if (restParam.isDefined) (c, t.init, Some(t.last))
-      else (c, t, None)
+
+    val (allNewParamsWithUsage, newBody) =
+      transformIsolatedBody(None, thisType, captureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
+
+    val captureParamCount = captureParams.size
+
+    val (newParams, newRestParam) = {
+      val t = allNewParamsWithUsage.iterator.drop(captureParamCount).map(_._1).toList
+      if (restParam.isDefined) (t.init, Some(t.last))
+      else (t, None)
     }
 
-    Closure(arrow, newCaptureParams, newParams, newRestParam, newBody, newCaptureValues)
+    val (newCaptureParams, newCaptureValues, unusedCaptureValues) = {
+      val newCaptureParams = List.newBuilder[ParamDef]
+      val newCaptureValues = List.newBuilder[Tree]
+      val unusedCaptureValues = List.newBuilder[Binding]
+      var unused = 0
```

Comment:
I know you said to avoid mutability but for this blob here, it was the easiest way to write it ☺️